sound/soc/sof/ipc4-loader.c

```diff
@@ -272,8 +272,85 @@ static int sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
 	return sof_ipc4_fw_parse_ext_man(sdev, plat_data->fw, false, 0);
 }
 
+static int sof_ipc4_load_library(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
+{
+	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+	struct snd_sof_pdata *plat_data = sdev->pdata;
+	const guid_t *widget_uuid = &swidget->uuid;
+	const char *fw_filename;
+	size_t fw_offset;
+	int ret, i;
+
+	/* nothing to do if widget UUID is not set */
+	if (!guid_is_null(&swidget->uuid))
+		return 0;
+
+	/* nothing to do if module library is already loaded */
+	for (i = 0; i < ipc4_data->num_fw_modules; i++)
+		if (guid_equal(&ipc4_data->module_uuids[i], widget_uuid))
+			return 0;
+
+	/*
+	 * The required 3rd party module libraries are dictated by the topology. So they should be
+	 * bundled in the same folder along with the topology binary
+	 */
+	fw_filename = kasprintf(GFP_KERNEL, "%s/%pUL.bin",
+				plat_data->tplg_filename_prefix, widget_uuid);
+	if (!fw_filename)
+		return -ENOMEM;
+
+	/* Index 0 is reserved for base firmware */
+	for (i = 1; i < ipc4_data->max_fw_libs; i++)
+		if (!ipc4_data->module_lib_info[i].id)
+			break;
+
+	if (i >= ipc4_data->max_fw_libs) {
+		dev_err(sdev->dev, "Library ID: %u exceeds max value %u\n",
+			i, ipc4_data->max_fw_libs);
+		ret = -EBUSY;
+		goto err;
+	}
+
+	ipc4_data->module_lib_info[i].id = i;
+	swidget->lib_info = &ipc4_data->module_lib_info[i];
+
+	ret = request_firmware(&ipc4_data->module_lib_info[i].fw, fw_filename, sdev->dev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "Library file '%s' not found\n", fw_filename);
+		goto err;
+	}
+
+	fw_offset = sof_ipc4_fw_parse_ext_man(sdev, ipc4_data->module_lib_info[i].fw, false, i);
+	if (fw_offset < 0)
+		goto release;
+
+	ipc4_data->module_lib_info[i].fw_offset = fw_offset;
+
+	if (sof_ops(sdev)->load_library) {
+		ret = sof_ops(sdev)->load_library(sdev, &ipc4_data->module_lib_info[i]);
+		if (ret < 0)
+			goto free;
+	}
```

Comment:
Humm, I don't get how this works. Is this not an error if the load_library is not supported at the chip level but explicitly referenced in the topology?

Better to fail than keep going IMHO.

Reply:
@plbossart IPC3 has no module library support. If I return an error here, I'd be forced to add a stub in the IPC3 ops.

Reply:
We may be talking on different levels here. 

My first point what that we could test if (sof_ops(sdev->load_library)) earlier and avoid requesting stuff that we know will not be supported, i.e. move the test to line 287.

The second point is what to do if the topology requests something that cannot be supported. That's an error, no?

Reply:
> @plbossart IPC3 has no module library support. If I return an error here, I'd be forced to add a stub in the IPC3 ops.

`sof_ipc4_load_library` will never be called by IPC3, right?

Reply:
ignore my comment please. I confused the comment to be for the load_library call in topology.c